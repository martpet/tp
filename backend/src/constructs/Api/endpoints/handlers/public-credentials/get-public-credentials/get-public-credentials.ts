import {
  APIGatewayProxyHandlerV2,
  CognitoIdentityClient,
  EnvVars,
  errorResponse,
  GetCredentialsForIdentityCommand,
  GetIdCommand,
  GetPublicCredentialsResponse,
} from 'lambda-layer';

const client = new CognitoIdentityClient({});

export const handler: APIGatewayProxyHandlerV2<
  GetPublicCredentialsResponse
> = async () => {
  const { identityPoolId } = process.env as EnvVars<'/public-credentials', 'GET'>;

  if (!identityPoolId) {
    return errorResponse('c6d852203f');
  }

  const { IdentityId } = await client.send(
    new GetIdCommand({ IdentityPoolId: identityPoolId })
  );

  if (!IdentityId) {
    return errorResponse('b84787e3e6');
  }

  const { Credentials } = await client.send(
    new GetCredentialsForIdentityCommand({ IdentityId })
  );

  if (!Credentials) {
    return errorResponse('d76fddbcb4');
  }

  return {
    access_key: Credentials.AccessKeyId,
    secret_key: Credentials.SecretKey,
    session_token: Credentials.SessionToken,
  };
};
