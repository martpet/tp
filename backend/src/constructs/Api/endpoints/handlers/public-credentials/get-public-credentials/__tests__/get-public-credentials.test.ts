import {
  CognitoIdentityClient,
  GetCredentialsForIdentityCommand,
  GetIdCommand,
} from '@aws-sdk/client-cognito-identity';
import { mockClient } from 'aws-sdk-client-mock';

import {
  itHasEnvVars,
  itResolves,
  itResolvesWithError,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';

import { handler } from '../get-public-credentials';

vi.mock('~/constructs/Api/utils/errorResponse');

process.env.identityPoolId = 'dummyIdentityPoolId';

const cognitoClientMock = mockClient(CognitoIdentityClient);

const args = [] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  cognitoClientMock.reset();

  cognitoClientMock.on(GetIdCommand).resolves({
    IdentityId: 'dummyIdentityId',
  });
  cognitoClientMock.on(GetCredentialsForIdentityCommand).resolves({
    Credentials: {
      AccessKeyId: 'dummyAccessKeyId',
      Expiration: new Date(),
      SecretKey: 'dummySecretKey',
      SessionToken: 'dummySessionToken',
    },
  });
});

describe('get-public-credentials', () => {
  itHasEnvVars(['identityPoolId'], handler, args);
  itSendsAwsCommand(GetIdCommand, cognitoClientMock, handler, args);
  itSendsAwsCommand(GetCredentialsForIdentityCommand, cognitoClientMock, handler, args);
  itResolves(handler, args);

  describe('when "IdentityPoolId" is missing from GetIdCommand response', () => {
    beforeEach(() => {
      cognitoClientMock.on(GetIdCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });

  describe('when "Credentials" is missing from GetCredentialsForIdentityCommand response', () => {
    beforeEach(() => {
      cognitoClientMock.on(GetCredentialsForIdentityCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });
});
