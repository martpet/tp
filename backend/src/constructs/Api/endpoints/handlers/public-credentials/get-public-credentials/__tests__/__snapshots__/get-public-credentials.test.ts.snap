// Vitest Snapshot v1

exports[`get-public-credentials > resolves with a correct value 1`] = `
{
  "accessKeyId": "dummyAccessKeyId",
  "expiration": 2022-01-01T00:00:00.000Z,
  "secretKey": "dummySecretKey",
  "sessionToken": "dummySessionToken",
}
`;

exports[`get-public-credentials > sends "GetCredentialsForIdentityCommand" from "CognitoIdentityClient" with correct args 1`] = `
{
  "IdentityId": "dummyIdentityId",
}
`;

exports[`get-public-credentials > sends "GetIdCommand" from "CognitoIdentityClient" with correct args 1`] = `
{
  "IdentityPoolId": "dummyIdentityPoolId",
}
`;

exports[`get-public-credentials > when "Credentials" is missing from GetCredentialsForIdentityCommand response > calls "errorResponse" with correct args 1`] = `
[
  [
    "d76fddbcb4",
  ],
]
`;

exports[`get-public-credentials > when "Credentials" is missing from GetCredentialsForIdentityCommand response > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-public-credentials > when "IdentityPoolId" is missing from GetIdCommand response > calls "errorResponse" with correct args 1`] = `
[
  [
    "b84787e3e6",
  ],
]
`;

exports[`get-public-credentials > when "IdentityPoolId" is missing from GetIdCommand response > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-public-credentials > when "identityPoolId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "c6d852203f",
  ],
]
`;

exports[`get-public-credentials > when "identityPoolId" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;
